db/commitlog/commitlog.cc

```diff
@@ -973,11 +986,22 @@ class db::commitlog::segment : public enable_shared_from_this<segment>, public c
             overhead += descriptor_header_size;
         }
```

Comment:
Should we write the header with the id in a separate op?

So we can have: write segment header, commit, rename, write buffers.

This prevents us from seeing a recycled buffer as containing valid data, and allows us to see a segment with zero entries (if we crash after activating the segment but before committing the first entry).

Reply:
? A recycled segment is named differently, so no risk of making mistakes there. There is a short time between discarding and rename done that we could pick up an old segment on next start, but that it fully valid as replaying _more_ is not an error. 

The recycled segment is then renamed before (re-)writing the first block, so if we crash after re-re-naming (now no longer recycled-commitlog), the readed will first read the file name, then the old header, see they don't match, and ignore the segment. 

I see no hole here...

Reply:
There's a hole in that if we replay an old segment, we can resurrect data. We're not going to fix it here, but if we strengthen commitlog so it can tell us when a segment is definitely NOT going to be replayed, we can use the new guarantee to fix that hole.

ref: https://github.com/scylladb/scylladb/pull/15524#discussion_r1338348669

Reply:
Scenario:

1. write data to owned token range
2. lose the token range (losing the sstables)
3. on another node, a tombstone is written and compacted away with the data
4. regain the token range
5. crash, replay commitlog
6. old commitlog for this range resurrects data

Having a separate write header step allows us to fence off complete commitlog segments.

Reply:
How is that more efficient than the file rename we do? An active segment is called "Commitlog-<blabla>...". Once dropped, it is renamed to "Recycled-Commitlog-<blabla>". We never pick up recycled files for replay. Ok, need a dir sync for file name maybe, but it seems not much different than doing the same + extra sector writes.

Reply:
Also, where in your example did we complete the segment? It is just as likely we have non-completed segments in this case, so if we should handle anything, it is general resurrection handling. Which is a different set of PR:s.

Reply:
The rename by itself is the source of the problem. It converts a segment which is no longer under consideration to being back to a replay candidate.

Wait, maybe I'm missing something - if the segment id in the header doesn't match the file name, we don't replay it. The segment id is repeated in the header, yes?

Reply:
Yes. Once the segment is renamed to something replayable, we will check the header against the file name, and if it does not match it is skipped.

Reply:
Ok. So so the rename is a logical zeroing of the segment.